### ISSUE TITLE
Gotify integration, integers in JSON PostAgent

### ISSUE THREAD

**Issue by stevelord** [huginn/huginn#2661]
Hi,

I'm trying to post alerts to Gotify and would like to dynamically specify a priority value. The [Gotify API](https://gotify.net/api-docs#/message/createMessage) expects one for creating a message, but I'm having a problem specifying one.

My Gotify Poster agent is set up as follows:

{
  "post_url": "https://my.gotify.box/message",
  "expected_receive_period_in_days": "365",
  "content_type": "json",
  "method": "post",
  "payload": {
    "title": "{{title}}",
    "message": "{{message}}",
    "priority": "{{priority}}"
  },
  "headers": {
    "X-Gotify-Key": "{% credential gotify_access_token %}"
  },
  "emit_events": "true",
  "output_mode": "merge",
  "no_merge": "false"
}

The credential definitely works. If I remove the priority the agent works, but then I can't use this for different classes of notification. If the priority is still set, it sends it as a string, but the Gotify API wants an integer. I had a look [here](https://github.com/huginn/huginn/issues/2352) but as_object didn't work for me. This is what I get back with the priority field set:

{
  "message": "This is a test message",
  "title": "Testing",
  "priority": "8",
  "body": "{\"error\":\"Bad Request\",\"errorCode\":400,\"errorDescription\":\"json: cannot unmarshal string into Go struct field MessageExternal.priority of type int\"}",
  "status": 400,
  "headers": {
    "Content-Length": "148",
    "Content-Type": "application/json",
    "Date": "Sat, 04 Jan 2020 12:35:10 GMT"
  }
}

Any thoughts on how I can ensure that priority is sent via the post as an integer, not a string?

**Comment by dsander** [comment]
Hi,

this should convert the string to a number: 

```
...
    "message": "{% assign message = message | plus: 0 %}",
...

```

**Comment by stevelord** [comment]
The variable I'm using is called ```priority```, so I swapped ```message``` for ```priority```. Unfortunately that doesn't look like it's changing it, this is what I get back:

[
  {
    "message": "This is a test message",
    "title": "Testing",
    "priority": "8",
    "body": "{\"error\":\"Bad Request\",\"errorCode\":400,\"errorDescription\":\"json: cannot unmarshal string into Go struct field MessageExternal.priority of type int\"}",
    "status": 400,
    "headers": {
      "Content-Length": "148",
      "Content-Type": "application/json",
      "Date": "Sun, 05 Jan 2020 13:11:04 GMT"
    }
  }
]

**Comment by dsander** [comment]
Woups, sorry about that, does this work?


```
...
    "priority": "{{priority | plus: 0 | as_object }}"
...
```

**Comment by stevelord** [comment]
Hmmm... It looks like there's something about the way the JSON is structured in the editor. I'm off to bed now but will try to take another look tomorrow and will report back once I have something.

**Comment by jarlave** [comment]
Press Toggle View and change `"priority": "8"` to `"priority": 8`

**Comment by dsander** [comment]
@jarlave I don't think that will work because the Agent configuration has to be valid JSON.

**Comment by jarlave** [comment]
I testet it now and you are right. It only works if you "hardcode" the number in without an liquid syntax.

**Comment by stevelord** [comment]
I found a solution and did more digging to try and find out what's going wrong. For anyone who's stuck, the thing to do is to create an agent that sets the priority value and message as normal strings. This can then be passed to a post agent that looks like this:

```
{
  "post_url": "https://my.gotify.box/message",
  "expected_receive_period_in_days": "365",
  "content_type": "json",
  "method": "post",
  "payload": {
    "title": "{{title}}",
    "message": "{{message}}",
    "priority": "{{priority | plus: 0 | as_object }}"
  },
  "headers": {
    "X-Gotify-Key": "{% credential gotify_access_token %}"
  },
  "emit_events": "true",
  "output_mode": "merge",
  "no_merge": "false"
}
```
The key part is the ```"priority": "{{priority | plus: 0 | as_object }}"```. It's not ideal but it works and this approach works in general for things that need integers.

From what I've been able to work out, the reason for this is that the text editor either adds quotes or needs quotes to validate the JSON.

I guess there are two reasonable fixes:

1. Support integers being added in the agent configuration editor.
2. Document the workaround so people don't trip up.

I would be inclined given discussions I've seen about a new version of Huginn, to suggest documenting the workaround and adding to the wiki. I'm happy to do so if that helps. Least I can do. Also I'm not good enough at Ruby or JSON to add integer support ;)

**Comment by dsander** [comment]
@stevelord Adding documentation about this to the Wiki would be great! 

We added the `as_object` "hack" because `Liquid` was designed as a HTML template engine and we are kind of abusing it. Since the JSON syntax is as it is and `Liquid`s default is to return strings, I don't think we have a lot of other options to enable use cases like this.

**Comment by OdinVex** [comment]
I've come across this issue and Gotify seems to prefer to not parse strings as integers, so it's become a real issue for my self. Using `as_object` hasn't helped, it's still a string.
